src/sentry/api/paginator.py

```diff
@@ -220,3 +223,95 @@ def get_result(self, limit=100, cursor=None):
             next=next_cursor,
             prev=prev_cursor,
         )
+
+
+def reverse_bisect_left(a, x, lo=0, hi=None):
+    """\
+    Similar to ``bisect.bisect_left``, but expects the data in the array ``a``
+    to be provided in descending order, rather than the ascending order assumed
+    by ``bisect_left``.
+
+    The returned index ``i`` partitions the array ``a`` into two halves so that:
+
+    - left side: ``all(val > x for val in a[lo:i])``
+    - right side: ``all(val <= x for val in a[i:hi])``
+    """
+    if lo < 0:
+        raise ValueError('lo must be non-negative')
+
+    if hi is None:
+        hi = len(a)
+
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if a[mid] > x:
+            lo = mid + 1
+        else:
+            hi = mid
+
+    return lo
+
+
+class SequencePaginator(object):
+    def __init__(self, data, reverse=False):
+        self.scores, self.values = map(
+            list,
+            zip(*sorted(data, reverse=reverse)),
+        ) if data else ([], [])
+        self.reverse = reverse
+        self.search = functools.partial(
+            bisect.bisect_left if not reverse else reverse_bisect_left,
+            self.scores,
+        )
+
+    def get_result(self, limit, cursor=None):
+        if cursor is None:
+            cursor = Cursor(0, 0, False)
+
+        assert cursor.offset > -1
+
+        if cursor.value == 0:
+            position = 0 if not cursor.is_prev else len(self.scores)
+        else:
+            position = self.search(cursor.value)
+
+        position = position + cursor.offset
+
+        if not cursor.is_prev:
```

Comment:
Would prefer these branches are swapped, so we have `if cursor.is_prev: ... else:` rather than `if not cursor.is_prev: ... else:`

Reply:
Can do, I was prioritizing the "normal case" of forward paging vs. reverse paging but I am find with making this change.

Reply:
maybe I just have PTSD from ruby `unless/else` blocks

Reply:
Flipped these here: 3321b763828c9dfd7a9e33434766245eca8a5d53